stablehlo/dialect/AssemblyFormat.cpp

```diff
@@ -371,5 +368,28 @@ ParseResult parseCustomCallTarget(AsmParser& parser, StringAttr& target) {
   return parser.parseSymbolName(target);
 }
 
+void printTypeExtensions(BoundedAttrInterface attr, DialectAsmPrinter& os) {
+  os << "bounds<";
+  auto printBound = [&](int64_t bound) { os << dimSizeToString(bound); };
+  llvm::interleaveComma(attr.getBounds(), os, printBound);
+  os << ">";
+}
+
+Attribute parseTypeExtensions(HloDialectInterface* dialect,
+                              DialectAsmParser& parser) {
+  SmallVector<int64_t> bounds;
+  auto parseBound = [&]() -> ParseResult {
+    if (!parser.parseOptionalQuestion()) {
```

Comment:
`succeeded(parser.parseOptionalQuestion())`?

Reply:
Done, also refactored the code that this was copypasted from.